Add Grading model and relation to Item (#11)

yarchick-balaban opened the pull request:
- Add GradingCompany model with certificate URL patterns
- Add GradingInfo model for item certification
- Create migration for grading_companies and grading_info tables
- Implement certificate URL generation
- Add grading info CRUD endpoints to items API
- Add fixtures for models according to models

Closes #11

Commit:
Add Grading model and relation to Item (#11)

- Add GradingCompany model with certificate URL patterns
- Add GradingInfo model for item certification
- Create migration for grading_companies and grading_info tables
- Add grading info CRUD endpoints to items API
- Add fixtures for models according to models for testings.
- Updated migrations for applying base grade data.

Grades and grade varitions can be added into Grading company:
All supported grades contains in grading company.

topol-3000 commented on backend/src/utils/certificate_url.py:
```diff
@@ -0,0 +1,282 @@
+            cert_number = cert_number.split("/")[-1]  # Take the part after the last slash
+
+    # Normalize grade for URL usage: remove + and * symbols, encode spaces
+    def normalize_grade_for_url(g: str) -> str:
```
why do not make it as a separate function?

topol-3000 commented on backend/src/utils/certificate_url.py:
```diff
@@ -0,0 +1,282 @@
+
+    # Check for special NGC categories
+    if normalized in ["NGC DETAILS", "NGCDETAILS"]:
+        return "NGCDetails"
```
a lot of magic strings that should be handled as enum or constants with description

topol-3000 commented on backend/src/utils/certificate_url.py:
```diff
@@ -0,0 +1,282 @@
+        return ""
+
+    # Normalize input - remove modifiers and extra spaces
+    normalized = grade.replace("+", "").replace("★", "").replace("*", "").strip().upper()
```
magic))
We should have a list(constant or tuple) of symbols to drop and then just remove them

topol-3000 commented on backend/src/utils/certificate_url.py:
```diff
@@ -0,0 +1,282 @@
+
+    norm_grade = normalize_grade_for_url(grade)
+
+    if company.short_name == "NGC":
```
the same situation with magic strings

topol-3000 commented on backend/src/schemas/item.py:
```diff
@@ -21,6 +22,21 @@ class ItemCreate(ItemBase):
     purchase_price: Annotated[int, Field(ge=0, description="Purchase price in pennies/cents")]
     purchase_date: Annotated[date | None, Field(description="Date when the item was purchased")] = None
 
+    # Required grading info fields for simple creation (like price history)
```
extra comment

topol-3000 commented on backend/src/models/grading_company.py:
```diff
@@ -0,0 +1,46 @@
+    is_active: Mapped[bool] = mapped_column(
+        Boolean, nullable=False, default=True, comment="Whether company is currently active"
+    )
+    grades: Mapped[dict[str, list[str]] | None] = mapped_column(
```
Why not a separate table for grades?

topol-3000 commented on backend/src/models/grading_info.py:
```diff
@@ -0,0 +1,48 @@
+    certificate_number: Mapped[str] = mapped_column(
+        String(255), nullable=False, comment="Unique certificate number from grading company"
+    )
+    grade: Mapped[str | None] = mapped_column(String(50), nullable=True, comment="Assigned grade (e.g., MS70, AU58)")
```
I'm a bit confused
we have grades in GradingCompany and here we have grade as string

yarchick-balaban replied:
Yes, in grade info can be one string as grade like "MS 65". Grades in Grading companies contains all supported grades for each company

topol-3000 commented on backend/src/api/routes/grading_company.py:
```diff
@@ -0,0 +1,143 @@
+from schemas.grading_company import GradingCompanyCreate, GradingCompanyRead, GradingCompanyUpdate
+from utils.certificate_url import generate_certificate_url
+
+router = APIRouter(prefix="/grading-companies", tags=["Grading Companies"])
```
just `tags=["Grading"]` ?

yarchick-balaban replied:
No, because we have grading info, just "grading" has logic relate's to info not to company

topol-3000 commented on backend/src/api/routes/grading_company.py:
```diff
@@ -0,0 +1,143 @@
+    certificate_url = generate_certificate_url(company, cert_number, grade)
+
+    return {
+        "company_id": company_id,
```
we have company_id, cert_number and grade in request parameters, Why do we need them in the response?

topol-3000 commented on backend/src/api/routes/grading_company.py:
```diff
@@ -0,0 +1,143 @@
+    await session.commit()
+
+
+@router.get("/certificate-url/preview")
```
"/{company_id}/certificate-url/preview ?

topol-3000 commented on backend/src/api/routes/grading_company.py:
```diff
@@ -0,0 +1,143 @@
+    for _category, grade_list in grades_dict.items():
+        all_grades.extend(grade_list)
+
+    return {
```
the same question with 
```
  "company_id": company_id,
        "company_name": company.name,
        "company_short_name": company.short_name,
```

why do we need them in the response?

topol-3000 commented on backend/src/api/routes/items.py:
```diff
@@ -117,6 +136,28 @@ async def create_item(
     session.add(price_history)
 
+    # Create grading info (always required, like price history)
+    # Minimal validation: check if company exists
```
just general comment, here a lot of auto generated useless comments.
just ask AI agen not to add meaningless  comments

topol-3000 commented on backend/src/api/routes/items.py:
```diff
@@ -117,6 +136,28 @@ async def create_item(
+    # Minimal validation: check if company exists
+    company_result = await session.execute(select(GradingCompany).where(GradingCompany.id == grading_company_id))
+    company = company_result.scalar_one_or_none()
+    if not company:
```
what is this???

we just get an Item, why do you return HTTP_400_BAD_REQUEST if there is no company?

yarchick-balaban replied:
MVP requires support for only rated coins. If the system does not have a grading company and, accordingly, gradings and URL templates for constructing a certification URL, we will not be able to add this coin. 

Support for verification coins (certificates from related companies) is also very important, based on this, grading companies are important as entities that know which links and grades each grading company supports.

topol-3000 commented on backend/src/api/routes/items.py:
```diff
@@ -117,6 +136,28 @@ async def create_item(
+
+    # Generate certificate URL immediately if we have the data
+    grading_info.company = company
+    grading_info.certificate_url = grading_info.generate_certificate_url()
```
the same
we are getting an Item we should not generate anything here

yarchick-balaban replied:
You are right. Need create urls according to grading companies pattern url on Front-end site. It's good practices when client can see what exactly url system generate and edit if needed.

yarchick-balaban replied:
Or live it here and when client detect wrong url he can edid this when edited item.

topol-3000 commented on backend/src/api/routes/items.py:
```diff
@@ -275,6 +316,144 @@ async def delete_item_price_history(
+) -> GradingInfo:
+    """Add grading info to an item."""
+    # Check if grading info already exists (one-to-one relationship)
+    existing_result = await session.execute(select(GradingInfo).where(GradingInfo.item_id == item.id))
```
we do not need it
just handle IntegrityError on insert

topol-3000 commented on backend/src/api/routes/items.py:
```diff
@@ -275,6 +316,144 @@ async def delete_item_price_history(
+            detail="Item already has grading info. Use PATCH to update or DELETE first.",
+        )
+
+    # Minimal validation: check if company exists
```
do not add useless checks
you won't be able to insert data if there is no company with specified company id

topol-3000 commented on backend/src/api/routes/items.py:
```diff
@@ -275,6 +316,144 @@ async def delete_item_price_history(
+        if not company:
+            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Grading company not found")
+
+    # Update fields that were provided
```
a lot of useless comments

topol-3000 commented on backend/src/migrations/versions/2025_08_25_0200-4ac6fb400756_add_complete_grading_system.py:
```diff
@@ -0,0 +1,322 @@
+
+
+def get_grading_companies_data() -> list[dict]:
+    """Get grading companies fixture data."""
```
it's not a fixture.
please, I have rechecked best practices and it should be as a separate SQL script to insert data

Commit:
Add unified SQL generation tool for production grading data

Add production-ready SQL generator for grading system deployment:

- generate_production_grading_data.py: Creates INSERT statements for both
  grading companies (PCGS, NGC, ANACS, ICG, CAC) and their comprehensive
  grade scales in single unified script
- Generates UUIDs using str(uuid4()) for PostgreSQL compatibility
- Outputs single production_grading_data.sql file ready for deployment
- Complete documentation with usage examples and Docker workflow
- Self-contained generation - no database connection required

Features:
- Unified script generates complete dataset (5 companies + ~900 grades)
- Company-specific grade variations (NGC uses PF, PCGS uses PR)
- Comprehensive grade categories: modern, ancient, details, special
- Support for selective company generation
- Production deployment workflow support

Generated files are gitignored and ready for production database deployment.

topol-3000 commented on backend/src/api/routes/items.py:
```diff
@@ -275,6 +313,127 @@ async def delete_item_price_history(
+    return grading_info
+
+
+@grading_info_router.get("/", response_model=GradingInfoRead | None)
```
why do not you use item_id in the url?

topol-3000 replied:
and the same question for other endpoints below

topol-3000 commented on backend/src/models/grading_company.py:
```diff
@@ -0,0 +1,57 @@
+        back_populates="company",
+        cascade="all, delete-orphan",
+    )
+    grades_list: Mapped[list["Grade"]] = relationship(
```
grades_list -> grades
